### PR TITLE
QR Gram Schmidt

### PR DESCRIPTION
**Pull request by rjenc29**
New Feature
QR - couple of basic algos to open the account.

- [x] Code conforms to style guide
- [ ] Doctests
- [x] Unit-tests
- [ ] Documentation
- [x] New line item in [releases.md](releases.md)
- [x] Info added to `literature` directory

Bug Report

- [ ] Minimal repro in a test function

### CODE REVIEW

File: tests/linear_algebra/test_qr.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+from fastats.core.ast_transforms.convert_to_jit import convert_to_jit
+from fastats.linear_algebra import qr, qr_classical_gram_schmidt
+
+
+qr_jit = convert_to_jit(qr)
+qr_qr_classical_gram_schmidt_jit = convert_to_jit(qr_classical_gram_schmidt)
+
+
+class QRTestMixin:
```
[the remark this answers is not in the document]
Comment (dwillmer) [reply tests/linear_algebra/test_qr.py]: @codacy-bot - you have issues.

ignoring this report

File: tests/linear_algebra/test_qr.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        assert_allclose(Q.T @ Q, np.eye(n), atol=1e-10)
+
+    @staticmethod
+    def check_versus_expectations(Q, Q_expected, R, R_expected, A):
+        assert_allclose(Q, Q_expected)
+        assert_allclose(R, R_expected)
+        assert_allclose(Q @ R, A)
+
+    def test_ucla(self):
+        # http://www.seas.ucla.edu/~vandenbe/133A/lectures/qr.pdf
```
Comment (dwillmer) [inline tests/linear_algebra/test_qr.py]: Can we turn this line comment into a docstring comment + literature folder addition?

File: tests/linear_algebra/test_qr.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        R_expected = np.array([[2, 4, 2],
+                               [0, 2, 8],
+                               [0, 0, 4]])
+
+        Q, R = self.fn(A)
+        self.check_versus_expectations(Q, Q_expected, R, R_expected, A)
+        self.assert_orthonormal(Q)
+
+    def test_wikipedia(self):
+        # https://en.wikipedia.org/wiki/QR_decomposition
```
Comment (dwillmer) [inline tests/linear_algebra/test_qr.py]: same here.

File: tests/linear_algebra/test_qr.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        A = np.array([[12, -51, 4],
+                      [6, 167, -68],
+                      [-4, 24, -41]])
+
+        Q_expected = np.array([[6/7, -69/175, -58/175],
+                               [3/7, 158/175, 6/175],
+                               [-2/7, 6/35, -33/35]])
```
Comment (dwillmer) [inline tests/linear_algebra/test_qr.py:64]: Highly approve of keeping these as fractions in the test code, makes it easy to check.

File: tests/linear_algebra/test_qr.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        R_expected = np.array([[14, 21, -14],
+                               [0, 175, -70],
+                               [0, 0, 35]])
+
+        Q, R = self.fn(A)
+        self.check_versus_expectations(Q, Q_expected, R, R_expected, A)
+        self.assert_orthonormal(Q)
+
+
+class QRTests(QRTestMixin, TestCase):
```
[the remark this answers is not in the document]
Comment (dwillmer) [reply tests/linear_algebra/test_qr.py]: ignored

File: tests/linear_algebra/test_qr.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    def setUp(self):
+        self.fn = qr
+
+
+class QRClassicalGSTests(QRTestMixin, TestCase):
```
[the remark this answers is not in the document]
Comment (dwillmer) [reply tests/linear_algebra/test_qr.py]: ignored

File: tests/linear_algebra/test_qr.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+class QRJitTests(QRTestMixin, TestCase):
+
+    def setUp(self):
+        self.fn = qr_jit
+
+
+class QRClassicalGSJitTests(QRTestMixin, TestCase):
+
+    def setUp(self):
+        self.fn = qr_qr_classical_gram_schmidt_jit
```
Comment (dwillmer) [inline tests/linear_algebra/test_qr.py]: qr_qr ?

Comment (rjenc29) [reply tests/linear_algebra/test_qr.py]: oops, double-vision - will fix

File: tests/linear_algebra/test_qr.py (hunk 3 above)
Comment (rjenc29) [reply tests/linear_algebra/test_qr.py]: Will do

**Comment by pawroman** [comment]
Good stuff. :1st_place_medal:

**Comment by rjenc29** [comment]
Other than the unrelated test failures, is this PR ready to merge in its current state?

**Comment by dwillmer** [comment]
@rjenc29 - yep, definitely, looks good to me

**Comment by pawroman** [comment]
The tests should pass with #56 merged - merging this now!

**Pull request merged**